### DISCUSSION TITLE
Complex numbers - to polar and from polar form

### DISCUSSION THREAD

**Discussion by RudolfSch** [PaddiM8/kalk#135] [Q&A] [answered]
Big thanks for the awesome calculator!

I searched quite a bit and kalker looks almost ideal for alternating current calculations where complex numbers are used.
However, complex numbers often have to be entered in polar coordinates (phi can be either in radians or degrees). In many cases you also need a complex value to be printed in polar coordinates. I couldn't find that in your documentation...

Is this already possible and if so could you give us some examples? If not, can this be realized as a function? 
Could you give me any hints please?

Thanks again for your engagement in this great software.

**Comment by RudolfSch** [comment]
Okay I tried further and found some hidden gems! If x is a complex number like x=1+2i I can use following built in functions.
Im(x), Re(x), abs(x), arg(x) to get the real or imaginary part and the absolute value and argument... That's great because with these I can define these two helper functions, which convert from polar to cartesian and I couldn't find in the built in functions list:

>> cart(A, phirad) = A*cos(phirad)+A*i*sin(phirad)
>> cartdeg(A, phideg) = A*cos(phideg/180*pi)+A*i*sin(phideg/180*pi)

What I still don't know is how I could implement a function which returns the polar form of with just one call. I tried this, but my rust knowledge is unfortunately too poor:
>> pol(x) = abs(x), arg(x)
Unexpected token: 'Comma'.

I also tried to put these functions to the default.kalker file on macOS via vi ~/Library/Application\ Support/kalker/default.kalker with this content (pi only defined because I thought it was not defined at this point)

pi=3.14159265359
cart(A, phirad) = A*cos(phirad)+A*i*sin(phirad)
cartdeg(A, phideg) = A*cos(phideg/180*pi)+A*i*sin(phideg/180*pi)

 I ran into further problems
>> cart(1, pi/3)
Undefined variable: 'c'.

So I would appreciate further tipps to make it smooth. Thx again for everything you created so far!

PS: I wanted to write about complex numbers in detail so that other users can also benefit from it.

**Reply by PaddiM8** [reply, marked as answer]
Hello! If you want to return two values, you can do that by doing something like `(abs(x), arg(x))` (wrapping them in parenthesis). Then you can access an inner value with the syntax `value[[i]]` or `var_i`, for example:

```julia
pol(x) = (abs(x), arg(x))
v = pol(3 - 4i)
x = v_1
y = v_2

x = v[[1]]
y = v[[2]]
```

As for the other issue, there seems to be a problem in kalker with having function parameters with several letters (eg. `phirad`), but replacing it with `p` makes it work for me (this is something I'll have to fix in the future, but it is probably only an issue with function parameters, not regular variables).

`cart(A, p) = A*cos(p)+A*i*sin(p)`

It looks like you placed `default.kalker` in the right folder, so it should detect it. You can confirm that it finds it by defining some variable, eg. `a = 3`, and then write `a` in kalker to see if it returns 3.

**Reply by RudolfSch** [reply]
Thanks for the fast reply, now it works like a charm! I added these lines to the default.kalker file:

cart(A, r) = A*cos(r)+A*i*sin(r)
cartdeg(A, d) = A*cos(d/180*pi)+A*i*sin(d/180*pi)

pol(x) = (abs(x), arg(x))
poldeg(x) = (abs(x), arg(x)/pi*180)

I guess this could be useful for other users as well. May I also ask if there is the possibility to add comments in this file? I tried // and it didn't work to document the things for later.

**Reply by PaddiM8** [reply]
Yeah this could probably be useful for other people too. I might add some default functions for this even.

Comments aren't a thing, but I guess it would make sense to have!
